Prometheus: Stabilize OpenTelemetry Exemplar to Prometheus Exemplar (#4964)
open-telemetry/opentelemetry-specification: specification/compatibility/prometheus_and_openmetrics.md

If the metric name for monotonic Sum metric points does not end in a suffix of `_total` a suffix of `_total` SHOULD be added by default, otherwise the name MUST remain unchanged. Exporters SHOULD provide a configuration option to disable the addition of `_total` suffixes.
Monotonic Sum metric points with `StartTimeUnixNano` should export the `{name}_created` metric as well.

If Sum is converted to a Prometheus Counter, then `Exemplars` MUST be converted
as described in the [Exemplar Conversion](#exemplar-conversion) section.
Otherwise, `Exemplars` SHOULD be dropped. If the Prometheus protocol only

> supports a single exemplar on the Counter sample, a random exemplar SHOULD be

Author: Question for reviewers: This now differs from the histogram bucket conversion.  Here we take a random exemplar.  The histogram bucket takes the latest exemplar.  Should we always use the latest exemplar?

Author: ```suggestion
supports a single exemplar on the Counter sample, the latest exemplar SHOULD be
```

- A series of `{name}_bucket` metric points that contain all attributes of the histogram point recorded as labels.  Additionally, a label, denoted as `le` is added denoting the bucket boundary. The label's value is the stringified floating point value of bucket boundaries, ordered from lowest to highest. The value of each point is the sum of the count of all histogram buckets up to the boundary reported in the `le` label. These points will include a single exemplar that falls within `le` label and no other `le` labelled point.  The final bucket metric MUST have an `+Inf` threshold.
- Histograms with `StartTimeUnixNano` set should export the `{name}_created` metric as well.

> `Exemplars` are converted as described in the [Exemplars](#exemplars-1) section.

Reviewer: ```suggestion
`Exemplars` are converted as described in the [Exemplar Conversion](#exemplar-conversion) section.
```

If the metric name for monotonic Sum metric points does not end in a suffix of `_total` a suffix of `_total` SHOULD be added by default, otherwise the name MUST remain unchanged. Exporters SHOULD provide a configuration option to disable the addition of `_total` suffixes.
Monotonic Sum metric points with `StartTimeUnixNano` should export the `{name}_created` metric as well.

If Sum is converted to a Prometheus Counter, then `Exemplars` MUST be converted
as described in the [Exemplar Conversion](#exemplar-conversion) section.
Otherwise, `Exemplars` SHOULD be dropped. If the Prometheus protocol only
supports a single exemplar on the Counter sample, a random exemplar SHOULD be

> converted.

Author: ```suggestion
converted. This matches the behavior of prometheus client libraries, which is to keep the
latest exemplar for counter instruments.
```

If the metric name for monotonic Sum metric points does not end in a suffix of `_total` a suffix of `_total` SHOULD be added by default, otherwise the name MUST remain unchanged. Exporters SHOULD provide a configuration option to disable the addition of `_total` suffixes.
Monotonic Sum metric points with `StartTimeUnixNano` should export the `{name}_created` metric as well.

If Sum is converted to a Prometheus Counter, then `Exemplars` MUST be converted
as described in the [Exemplar Conversion](#exemplar-conversion) section.
Otherwise, `Exemplars` SHOULD be dropped. If the Prometheus protocol only
supports a single exemplar on the Counter sample, the latest exemplar SHOULD be

> converted. This matches the behavior of prometheus client libraries, which is to keep the

Author: ```suggestion
converted. This matches the behavior of Prometheus client libraries, which is to keep the
```

  result being that the Offset fields are different-by-one.
- `Min` and `Max` are not used.
- `StartTimeUnixNano` is not used.

> - `Exemplars` are converted as described in the [Exemplars](#exemplars-1) section.

Reviewer: ```suggestion
- `Exemplars` are converted as described in the [Exemplar Conversion](#exemplar-conversion) section.
```

- A single `{name}_count` metric denoting the count field of the histogram. All attributes of the histogram point are converted to Prometheus labels.
- `{name}_sum` metric denoting the sum field of the histogram, reported only if the sum is positive and monotonic. The sum is positive and monotonic when all buckets are positive. All attributes of the histogram point are converted to Prometheus labels.
- A series of `{name}_bucket` metric points that contain all attributes of the histogram point recorded as labels.  Additionally, a label, denoted as `le` is added denoting the bucket boundary. The label's value is the stringified floating point value of bucket boundaries, ordered from lowest to highest. The value of each point is the sum of the count of all histogram buckets up to the boundary reported in the `le` label.  The final bucket metric MUST have an `+Inf` threshold.
- Histograms with `StartTimeUnixNano` set should export the `{name}_created` metric as well.

`Exemplars` are converted as described in the [Exemplar Conversion](#exemplar-conversion) section.
If the Prometheus protocol only supports a single exemplar per-bucket, the latest

> exemplar that falls into each bucket MUST be converted.

Reviewer: Should the verb be "SHOULD", to match the Sum language on line 499?
```suggestion
exemplar that falls into each bucket SHOULD be converted.
```

the same Prometheus key. In such cases, the values MUST be concatenated together,
separated by `;`, and ordered by the lexicographical order of the original keys.

### Exemplar Conversion

**Status**: [Stable](../document-status.md)

> [OpenTelemetry Exemplars](../metrics/data-model.md#exemplars) MUST be converted

Reviewer: I suggest the following to disambiguate:
```suggestion
When an exemplar is converted per the metric-type-specific sections above,
the [OpenTelemetry Exemplar](../metrics/data-model.md#exemplars) MUST be converted
```

to Prometheus exemplars if the Prometheus (push or pull) protocol being used

> supports them as follows:

Reviewer: Disambiguate:
```suggestion
supports them, as follows:
```

Author: That doesn't quite read correctly to me.

Reviewer: @dashpole Without the comma, IMO it reads as "if the Prometheus (push or pull) protocol being used supports them as follows", i.e. it could look as if the "as follows" refers to the protocol.

> supports them as follows:

Author: ```suggestion
to a Prometheus exemplar if the Prometheus (push or pull) protocol being used
supports them as follows:
```
